token-group/interface/src/instruction.rs

```diff
@@ -0,0 +1,495 @@
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:initialize_group")]
+pub struct InitializeGroup<G>
+where
+    G: SplTokenGroup,
+{
+    /// Update authority for the group
+    pub update_authority: OptionalNonZeroPubkey,
+    /// The maximum number of group members
+    pub max_size: Option<u64>,
+    /// Additional state
+    pub meta: Option<G>,
+}
+
+/// Instruction data for updating the max size of a `Group`
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:update_group_max_size")]
+pub struct UpdateGroupMaxSize {
+    /// New max size for the group
+    pub max_size: Option<u64>,
+}
+
+/// Instruction data for updating the authority of a `Group`
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:update_group_authority")]
+pub struct UpdateGroupAuthority {
+    /// New authority for the group, or unset if `None`
+    pub new_authority: OptionalNonZeroPubkey,
+}
+
+/// Instruction data for initializing a new `Member` of a `Group`
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:initialize_member")]
+pub struct InitializeMember {
+    /// The pubkey of the `Group`
+    pub group: Pubkey,
+    /// The member number
+    pub member_number: u64,
+}
+
+/// The type of item to emit
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize)]
+pub enum ItemType {
+    /// Emit the `Group`
+    Group,
+    /// Emit the `Member`
+    Member,
+}
+
+/// Instruction data for `Emit`
+#[derive(Clone, Debug, PartialEq, BorshSerialize, BorshDeserialize, SplDiscriminate)]
+#[discriminator_hash_input("spl_token_group_interface:emitter")]
+pub struct Emit {
+    /// Start of range of data to emit
+    pub start: Option<u64>,
+    /// End of range of data to emit
+    pub end: Option<u64>,
+    /// The type of item to emit
+    pub item_type: ItemType,
+}
+
+/// All instructions that must be implemented in the SPL Token Group Interface
+///
+/// Note: Any instruction can be extended using additional required accounts by
+/// using the `InitializeExtraAccountMetaList` instruction to write
+/// configurations for extra required accounts into validation data
+/// corresponding to an instruction's unique discriminator.
+#[derive(Clone, Debug, PartialEq)]
+pub enum TokenGroupInterfaceInstruction<G>
+where
+    G: SplTokenGroup,
+{
+    /// Initialize a new `Group`
+    ///
+    /// Assumes one has already initialized a mint for the
+    /// group.
+    ///
+    /// Accounts expected by this instruction:
+    ///
+    ///   0. `[w]`  Group
+    ///   1. `[]`   Mint
+    ///   2. `[s]`  Mint authority
+    ///   3..3+M `[]` `M` additional accounts, written in validation account
+    /// data
```

Comment:
Maybe I'm mistakenly pegging this concept to the interface itself, but if you check out the processor functions for creating an `Original` and `Reprint`, they are using extra metas to handle some of the metadata work, which again is specific to this example.

The idea is - technically - you can use extra metas to build on _any_ interface instruction, and I believe that should be the case for any interface instruction, not just transfer hook's "execute".

The idea of extra metas being a more universal thing for interfaces kind of builds on the other topic of making it it's own interface. wdyt?

Reply:
Ah I think I see what you mean. Rather than complicating any particular interface, how about creating a client library that initializes everything through multiple different interfaces?

For example, if you want to create an `Original`, you'll:

* create the mint (token interface)
* create the metadata (metadata interface)
* create the collection (collection interface)

and for reprints, you'll:

* create a mint (token interface)
* copy the metadata from the original, creating more new metadata (metadata interface)
* create the member account (collection interface)

Reply:
Yeah, I mean this makes perfect sense, but I believe there will be folks who want this all to be handled within one instruction on their program, and likely add other guards/business logic to the steps demonstrated.

For our example, we can go with this approach. I felt like it was important to demonstrate how interfaces can stack. But maybe that demonstration can live elsewhere?

Reply:
We can always have a separate program, a stateless protocol, that just CPIs into all of these instructions in one.

The example that implements multiple interfaces is definitely valuable, so we should definitely have a program somewhere that implements collection + metadata. People will likely use that as a template more than either separate example

Reply:
Ahh yeah, a stateless example is probably a good route to take